sigstore/sign.py

```diff
@@ -89,86 +89,105 @@ class Signer:
     The primary API for signing operations.
     """
 
-    def __init__(self, identity_token: str, cache: bool = True) -> None:
+    def __init__(
+        self, identity_token: str, fulcio: FulcioClient, cache: bool = True
+    ) -> None:
         """
         Create a new `Signer`.
 
         `identity_token` is the identity token used to request a signing certificate
         from Fulcio.
 
+        `fulcio` is a `FulcioClient` capable of connecting to a Fulcio instance
+        and returning signing certificates.
+
         `cache` determines whether the signing certificate and ephemeral private key
         should be reused (until the certificate expires) to sign different artifacts.
         Default is `True`.
         """
         self._identity_token: str = identity_token
-        self.cache: bool = cache
-        self._signing_certificate: Optional[FulcioCertificateSigningResponse] = None
-        self._private_key: Optional[ec.EllipticCurvePrivateKey] = None
+        self.__cached_private_key: Optional[ec.EllipticCurvePrivateKey] = None
+        self.__private_key: Optional[ec.EllipticCurvePrivateKey] = None
+        self.__cached_signing_certificate: Optional[
+            FulcioCertificateSigningResponse
+        ] = None
+        self.__signing_certificate: Optional[FulcioCertificateSigningResponse] = None
+        if cache:
+            logger.debug("Generating ephemeral keys...")
+            self.__cached_private_key = ec.generate_private_key(ec.SECP384R1())
+            logger.debug("Requesting ephemeral certificate...")
+            self.__cached_signing_certificate = self._signing_cert(
+                self._private_key, fulcio
+            )
+
+    def __del__(self) -> None:
+        """
+        Delete the Signer internal attributes.
+        This allows to not persist the certificate and private key
+        outside of the scope defined by `SigningContext`.
+        """
+        del self.__signing_certificate
+        del self.__private_key
 
     @property
-    def private_key(self) -> ec.EllipticCurvePrivateKey:
+    def _private_key(self) -> ec.EllipticCurvePrivateKey:
         """Get or generate a signing key."""
-        if not self.cache or self._private_key is None:
+        if self.__cached_private_key is None:
             logger.debug("Generating ephemeral keys...")
-            self._private_key = ec.generate_private_key(ec.SECP384R1())
+            self.__private_key = ec.generate_private_key(ec.SECP384R1())
```

Comment:
Per above: rather than storing into `__private_key`, I think we can return directly here.